src/cloud-api-adaptor/docs/SecureComms.md

```diff
@@ -0,0 +1,99 @@
+```sh
+ExecStart=/usr/local/bin/agent-protocol-forwarder -kata-agent-namespace /run/netns/podns -secure-comms -kata-agent-socket /run/kata-containers/agent.sock $TLS_OPTIONS $OPTIONS
+```
+
+You may also include additional Inbounds and Outbounds configurations to the Forwarder using the `-secure-comms-inbounds` and `-secure-comms-outbounds` flags. See more details regarding Inbounds and Outbounds below.
+
+Once you changed `podvm/files/etc/systemd/system/agent-protocol-forwarder.service`, you will need to [rebuild the podvm](./../podvm/README.md).
+
+
+### Activate CAA Secure-Comms feature
+
+Add to the `peer-pods-cm` config map at the `confidential-containers-system` namespace:
+```sh
+apiVersion: v1
+data:
+    ...
+    SECURE_COMMS: "true"
+    ...
+```
+
+You may also include additional Inbounds and Outbounds configurations to the Adaptor using the `SECURE_COMMS_INBOUNDS` and `SECURE_COMMS_OUTBOUNDS` config points. See more details regarding Inbounds and Outbounds below.
+
+You may also set the KBS address using the `SECURE_COMMS_KBS_ADDR` config point.
+
+
+### Adding named tunnels to the SSH channel
+Named tunnels can be added to the SSH channel. Adding a named tunnel requires adding an Inbound at one of the SSH channel peers and an Outbound at the other SSH channel peer. The Inbound and Outbound both carry the name of the tunnel being created.
+
+        |---------Tunnel----------| 
+Client->Inbound----------->Outbound->Server
+
+
+Inbounds and Outbounds take the form of a comma separated inbound/outbound tags such that Inbounds are formed as "InboundTag1,InboundTag2,InboundTag3,..." and Outbounds are formed as "OutboundTag1,OutboundTag2,outboundTag3,..."
+
+Each Inbound tag is structured as `Phase:Name:Port` where:
+- Phase can be 'K' to represent an inbound available during the Kubernetes phase, 'A' to represent an inbound available during the Attestation phase, or 'B' to represent an inbound available during both phases.
+- Name is the name of the tunnel
+- Port is the local service port being opened to serve as ingress of the tunnel.
+
+Each outbound tag is structured as `Phase:Name:Host:Port` or `Phase:Name:Port` where:
+- Phase can be 'K' to represent an outbound available during the Kubernetes phase, 'A' to represent an outbound available during the Attestation phase, or 'B' to represent an outbound available during both phases.
+- Name is the name of the tunnel
+- Host (if available) is a hostname or IP address of the destination host to forward the information to.
+- Port is the destination port to forward the information to.
+
+For example, an outbound tag such as `K:ABC:myhost.com:1234` means that during the `Kubernetes phase`, an output of a tunnel named `ABC` is registered, such that information from a client connecting to ABC Inbound will be tunneled and forwarded to `myhost.com` port `1234`).
+
+## Testing
+
+Testing securecomms as a standalone can be done by using:
```

Comment:
Sorry, ignorant question here about what standalone means given that it is using kubeconfig? I also tried to run it and got:
```
# go run ./test/securecomms/double/main.go
go: downloading golang.org/x/oauth2 v0.10.0
go: downloading github.com/evanphx/json-patch v4.12.0+incompatible
go: downloading google.golang.org/protobuf v1.31.0
2024/04/24 05:34:42 [secure-comms] Using PP SecureComms: InitSshServer version v0.2
2024/04/24 05:34:42 [secure-comms] Inbound listening to port 7000
2024/04/24 05:34:42 [secure-comms] Inbound listening to port 16443
2024/04/24 05:34:42 [secure-comms] Inbound listening to port 9053
2024/04/24 05:34:42 [secure-comms] SSH service starting on port: 2222
2024/04/24 05:34:42 [secure-comms] Attestation phase: waiting for client to connect
2024/04/24 05:34:42 [secure-comms] DeleteSecret 'pp-sid' error secrets "pp-sid" not found
2024/04/24 05:34:42 [secure-comms] Using PP SecureComms: InitSshClient version v0.2
2024/04/24 05:34:46 [secure-comms] CreateSecret 'sshclient'
2024/04/24 05:34:46 InitSshClient failed to read KBS client secret: secrets "kbs-client" not found
```
so does the KBS need to be deployed first to try this out?

Reply:
Just to note that with the cluster, kbs and secure comms set up I got this working:
```
# go run ./test/securecomms/double/main.go
2024/05/02 12:35:54 [secure-comms] Using PP SecureComms: InitSshServer version v0.2
2024/05/02 12:35:54 [secure-comms] Inbound listening to port 7000
2024/05/02 12:35:54 [secure-comms] Inbound listening to port 16443
2024/05/02 12:35:54 [secure-comms] Inbound listening to port 9053
...
2024/05/02 12:36:02 [secure-comms] Attestation phase: starting
2024/05/02 12:36:02 [secure-comms] Attestation phase: client connected
2024/05/02 12:36:02 [secure-comms] Attestation phase: connected
2024/05/02 12:36:02 [secure-comms] Attestation phase: ssh connected - 127.0.0.1:2222
2024/05/02 12:36:03 [secure-comms] Attestation phase: SSH server initialized keys
2024/05/02 12:36:03 [secure-comms] Attestation phase: SSH server initialized with NoClientAuth
2024/05/02 12:36:03 [secure-comms] Attestation phase: ssh skipped validating server's host key (type ssh-rsa) during attestation
2024/05/02 12:36:03 [secure-comms] Attestation phase: logged-in without key
2024/05/02 12:36:03 [secure-comms] Attestation phase: peer reported phase Attestation
2024/05/02 12:36:03 [secure-comms] Attestation phase: peer reported phase Attestation
...
2024/05/02 12:36:03 [secure-comms] Attestation phase: done
HttpClient start : http://127.0.0.1:45561
HttpClient sending req: http://127.0.0.1:45561
2024/05/02 12:36:03 [secure-comms] Kubernetes phase: starting (number of restarts 0)
2024/05/02 12:36:03 [secure-comms] Kubernetes phase: ssh connected - 127.0.0.1:2222
2024/05/02 12:36:03 [secure-comms] Kubernetes client connected
2024/05/02 12:36:03 [secure-comms] Kubernetes phase: connected
...
2024/05/02 12:36:04 [secure-comms] Kubernetes phase: done
2024/05/02 12:36:04 [secure-comms] SshClientInstance DisconnectPP success
2024/05/02 12:36:04 [secure-comms] DeleteSecret 'pp-sid'
*** SUCCESS ***
```
Maybe not what I'd describe as standalone as so, but it's good to get it working :)